gmcs/linglib/lexbase.py

```diff
@@ -166,13 +166,13 @@ def roots(self):
         hierarchy for this position class. Note that there may be more than
         one root for any given position class.
         """
-        return [n for n in list(self.nodes.values()) if len(n.parents()) == 0]
+        return [n for n in self.nodes.values() if len(n.parents()) == 0]
 
     def input_span(self):
         return self.hierarchy.get_lineage(key=self.key, relation='input')
 
     def valid_inputs(self):
-        all_inps = list(self.input_span().values())
+        all_inps = self.input_span().values()
```

Comment:
Here you probably want to keep the `list()`.

Reply:
Hmm I did run the tests. But I will put it back I guess.

Reply:
It's possible that the return value of `valid_inputs()` is only iterated over (or other operations that don't need a list), but it's better for our functions to return lists than view objects to avoid problems down the road.

Reply:
Sorry, I probably should have been more specific about the weird list thing! It’s specifically the `for x in list(d.keys())`/`for x in list(d.values())`. As @goodmami indicates, in Python 2, keys/values return a list, whereas in Python 3 they return special keys/values view objects, so just iterating over them is an unnecessary cast, but otherwise it’s probably good to keep them.